server/src/main/java/io/crate/analyze/relations/RelationAnalysisContext.java

```diff
@@ -72,22 +73,39 @@ void addJoinPair(JoinPair joinType) {
     }
 
     void addJoinType(JoinType joinType, @Nullable Symbol joinCondition) {
-        int size = sources.size();
-        assert size >= 2 : "sources must be added first, cannot add join type for only 1 source";
-        Iterator<RelationName> it = sources.keySet().iterator();
-        RelationName left = null;
-        RelationName right = null;
-        int idx = 0;
-        while (it.hasNext()) {
-            RelationName sourceName = it.next();
-            if (idx == size - 2) {
-                left = sourceName;
-            } else if (idx == size - 1) {
-                right = sourceName;
+        if (joinCondition == null) {
+            // If join condition is NULL, it's a CROSS join and we can just take
+            // item(size - 1): table being cross joined, right side of the join node.
+            // item(size - 2): last table in the join chain
```

Comment:
we keep the old logic for cross join, taking last and second last tables is fine in this case.